opentelemetry-sdk/src/metrics/internal/sum.rs

```diff
@@ -10,26 +11,39 @@ use opentelemetry::{global, metrics::MetricsError};
 
 use super::{
     aggregate::{is_under_cardinality_limit, STREAM_OVERFLOW_ATTRIBUTE_SET},
-    Number,
+    AtomicTracker, Number,
 };
 
 /// The storage for sums.
-#[derive(Default)]
 struct ValueMap<T: Number<T>> {
     values: Mutex<HashMap<AttributeSet, T>>,
+    has_no_value_attribute_value: AtomicBool,
+    no_attribute_value: T::AtomicTracker,
```

Comment:
Sorry, missed out on reviewing this. It seems  ValueMap instance will either store a HashMap or a single value (`no_attribute_value` along with `has_no_value_attribute_value`), but not both. Should we instead use enum to segregate them, for efficient memory utilization. For no-attributes scenario, even for empty `values` object, the ovehead from metdata for Mutex and HashMap would be more than that of enum. Ignore if this doesn't makes sense, still going through the implementation :)

Reply:
This is not true.  You can call Add with no attributes and then add again with an attribute set.  So one counter can have both

Reply:
@KallDrexx use case relies on user discipline to ensure no attributes are passed, not something SDK can assume and do optimizations!